core/src/main/scala/org/apache/spark/executor/ExecutorMetrics.scala

```diff
@@ -28,35 +30,63 @@ import org.apache.spark.metrics.ExecutorMetricType
 @DeveloperApi
 class ExecutorMetrics private[spark] extends Serializable {
 
-  // Metrics are indexed by MetricGetter.values
-  private val metrics = new Array[Long](ExecutorMetricType.values.length)
+  private var metrics = mutable.Map.empty[String, Long]
```

Comment:
There's a fair amount of overhead in storing values in a Map, data structure and the String keys. All this would then be sent in the heartbeat, so this would add overhead to the heartbeats. The array representation is a lot more compact. Unfortunately there's no EnumMap in Scala. I'd prefer to keep the array storage, and use map for returning and passing in values.

Reply:
Overhead is a valid concern. However, this isn't a huge overhead and it isn't always there as we have configs for whether user wants computing executor metrics. I can also make the map to be smaller by changing it from being Map[String, Long] to something like Map[somesmallertype, Long] and just having the mapping  to metric names when we are writing to or reading from json.
I used map since I saw a few comments here and also in [21221](https://github.com/apache/spark/pull/21221) about using map instead of array. We should definitely decide about this now before the next release since we can't make changes to the API after that. I'm open to moving back to using Array if it is really unwanted.

Reply:
I think I have found a better solution to use an array like before and make use of a map for clarity. I will try to push this new solution by Wednesday night after finishing the tests. Thank you very much.